### PR TITLE
Add Multipart Support to Web Cmdlets

### PR DESCRIPTION
**Pull request by markekraus**
Summary
----------
Partially implements  #2112

- Adds `System.Net.Http.MultipartFormDataContent` as a possible type for `-Body`
- Adds `/Multipart/` test to WebListener 

This allows for the user to create their own `Http.MultipartFormDataContent` object and submit it. Since `multipart/form-data` submissions are highly flexible, adding direct support for it to the CmdLets may over-complicate the command parameters and a limited implementation would not address the broad scope of use cases. This at least allows the user to submit multipart forms using the Web Cmdlets and not have to manage their own `HttpClient`. Once this is introduced, limited multipart implementations can be expanded to use the code in this PR.

Documentation Needed
--------------------------
Note that documentation will probably need to make clear that `-ContentType` will be ignored when a `MultipartFormDataContent` object is supplied to `-Body`. Similarly, content headers specified by `-Headers` and `-WebSession` will be ignored too.

`ContentType` must be set by the `MultipartFormDataContent` object as it controls the `boundary` in the `Content-Type` header used to distinguish between the various fields supplied. It doesn't make sense to use any other `Content-Type` header when doing `multipart/form-data` anyway.  If this isn't done here, the Foreach on line 421 throws. when `Content` is set with a `MultipartFormDataContent` it overwrites all the content headers, so trying to add content headers again causes an exception about `ContentType` already exists.

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: We should exclude double casting. Here it is better use C# 7.0 pattern:
```c#
else if (content is MultipartFormDataContent multipartFormDataContent ) 
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: We should use curly braces:
```c#
if ( ... )
{
    ...
}
```
or single line operator.

File: test/tools/WebListener/Controllers/MultipartController.cs
Comment (iSazonov) [inline test/tools/WebListener/Controllers/MultipartController.cs]: Please remove extra line.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: I can do this, but it will not match the surrounding code.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: Awesome. I will do that.

File: test/tools/WebListener/Controllers/MultipartController.cs
Comment (markekraus) [reply test/tools/WebListener/Controllers/MultipartController.cs]: will do

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: I believe we should use approved patterns for new code.
You can wait a view of other maintainers.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: We use the same Context header 😕 
I believe we should add cmdlet name `Context "Invoke-WebRequest Multipart/form-data Tests"`

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Context "Invoke-RestMethod Multipart/form-data Tests"

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (markekraus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Yes, because they are under different describe blocks. We did the same thing with HTTPS

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: I think it's somewhat confusing, but we can leave it for Context.

Comment (markekraus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: For clarity, would you like me to change it or leave it as is?

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Leave it as is.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: I agreed with @iSazonov. The new code is somewhat isolated as it's in its own method, so I think it's fine to use the more preferred style.

Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: @daxian-dbw Awesome! format is already corrected in the most recent commit for this PR.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: `WebSession.ContentHeader` is set at line 325 when `ContentType` is specified. Is it OK to ignore that as well?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: nit: can you please also mention why it's required -- `$multipartContent` will be unwrapped/enumerated otherwise.

**Review by daxian-dbw — COMMENTED** [review]
@markekraus Thank you for the good work! Could you please take a look at the 2 comments I left?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: Yes, `ContentType` must be set by the `MultipartFormDataContent` object as it controls the `boundary` in the `Content-Type` header used to distinguish between the various fields supplied. Documentation will probably need to make clear that `-ContentType` will be ignored when a `MultipartFormDataContent` object is supplied. It doesn't make sense to use any other `Content-Type` header when doing `multipart/form-data` anyway.  If this isn't done here, the Foreach on line 421 throws. when `Content` is set with a `MultipartFormDataContent` it overwrites all the content headers, so trying to add content headers again causes an exception about `ContentType` already existing.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (markekraus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: fixed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs]: Thanks for the clarification. I updated the PR description to capture your words.

Then there is potential problem at line 239:
```
if (WebSession.Headers.Count > 0)
{
    WebSession.ContentHeaders.Clear();
    foreach (var entry in WebSession.Headers)
    {
        if (HttpKnownHeaderNames.ContentHeaders.Contains(entry.Key))
        {
            WebSession.ContentHeaders.Add(entry.Key, entry.Value);
        }
...
```
If `-Headers` or `-WebSession` is specified, then the `foreach` on line 421 would throw.

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Review by iSazonov — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@markekraus Thanks again for the good work!

**Comment by markekraus** [comment]
@daxian-dbw For documentation, when is that handled? Should I submit a PR/Issue to PowerShell-Docs now or does it need to wait for the next beta release?

**Comment by daxian-dbw** [comment]
Yes, you are welcome to submit a PR to [PowerShell-Docs](https://github.com/PowerShell/PowerShell-Docs) now :)
Here is the guideline for maintainers about PRs that require doc changes (see [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)):
> If your change warrants an update to user-facing documentation, a Maintainer will add the Documentation Needed label to your PR and add an issue to the PowerShell-Docs repository, so that we make sure to update the official documentation to reflect your contribution. As an example, this requirement includes any changes to cmdlets (including cmdlet parameters) and features which have associated about_* topics. While not required, we appreciate any contributors who add this label and create the issue themselves. Even better, all contributors are free to contribute the documentation themselves. (See Contributing to documentation related to PowerShell for more info.)

Please see https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#contributing-to-documentation for more details

**Pull request merged**
